src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,205 @@
+        Node<T> nodeT = tail;
+        for (int i = 0; i < size - index - 1; i++) {
+            nodeT = nodeT.prev;
+        }
+        return nodeT;
+    }
+
+    public void unlink(Node node) {
+        Node previous = node.prev;
+        Node next = node.next;
+        if (previous == null && size != 1) {
+            next.prev = null;
+            head = next;
+        } else if (next == null && size != 1) {
+            previous.next = null;
+            tail = previous;
+        } else if (size == 1) {
+            head.prev = null;
+            head.next = null;
+            head.value = null;
+        } else {
+            previous.next = next;
+            next.prev = previous;
+        }
+    }
+
+    private void indexCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Invalid index");
+        }
+    }
+
+    private void indexCheckAdd(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Invalid index");
+        }
+    }
+
+    private void addByIndex(int index, T value) {
+        if (index == size && size != 0) {
+            insertLast(value);
+        } else if (index == 0 && size != 0) {
+            insertHead(value);
+        } else if (index == 0 && size == 0) {
+            insertFirst(value);
+        } else {
+            insert(iterator(index), value);
+        }
+    }
+
+    public boolean removeByValue(T object) {
+        Node<T> nodeT = head;
+        for (int i = 0; i < size; i++) {
+            if (nodeT.value == object || nodeT.value != null && nodeT.value.equals(object)) {
+                unlink(nodeT);
+                size--;
+                return true;
+            } else {
+                nodeT = nodeT.next;
+            }
+        }
+        return false;
+    }
+
+    private T removeByIndex(int index) {
+        Node old = iterator(index);
+        indexCheck(index);
+        unlink(old);
+        size--;
+        return (T)old.value;
+    }
+
+    private T setByIndex(int index, T value) {
+        indexCheck(index);
+        Node<T> nodeT = iterator(index);
+        T oldValue = nodeT.value;
+        nodeT.value = value;
+        return oldValue;
+    }
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            insertFirst(value);
+        } else {
+            insertLast(value);
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        indexCheckAdd(index);
+        addByIndex(index, value);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        return iterator(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        return setByIndex(index, value);
     }
```

Comment:
Why do you need a helper method setByIndex? Why would you need to create an additional method if the only thing your target method does is just to call that additional method? Just move the impl of setByIndex to the method set.